src/app/clusters/temperature-control-server/temperature-control-server.cpp

```diff
@@ -0,0 +1,161 @@
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+};
+
+} // namespace
+TemperatureControlAttrAccess gAttrAccess;
+
+CHIP_ERROR TemperatureControlAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == TemperatureControl::Id);
+
+    const TemperatureControl::SupportedTemperatureLevelsManager * gSupportedTemperatureLevelManager =
+        TemperatureControl::GetSupportedTemperatureLevelsManager();
+
+    if (TemperatureControl::Attributes::SupportedTemperatureLevels::Id == aPath.mAttributeId)
+    {
+        const TemperatureControl::SupportedTemperatureLevelsManager::TemperatureLevelOptionsProvider
+            temperatureLevelOptionsProvider =
+                gSupportedTemperatureLevelManager->GetTemperatureLevelOptionsProvider(aPath.mEndpointId);
+        if (temperatureLevelOptionsProvider.begin() == nullptr)
+        {
+            aEncoder.EncodeEmptyList();
+            return CHIP_NO_ERROR;
+        }
+        CHIP_ERROR err;
+        err = aEncoder.EncodeList([temperatureLevelOptionsProvider](const auto & encoder) -> CHIP_ERROR {
+            const auto * end = temperatureLevelOptionsProvider.end();
+            for (auto * it = temperatureLevelOptionsProvider.begin(); it != end; ++it)
+            {
+                auto & temperatureLevelOption = *it;
+                ReturnErrorOnFailure(encoder.Encode(temperatureLevelOption));
+            }
+            return CHIP_NO_ERROR;
+        });
+        ReturnErrorOnFailure(err);
+    }
+    return CHIP_NO_ERROR;
+}
+
+bool TemperatureControlHasFeature(EndpointId endpoint, TemperatureControl::Feature feature)
+{
+    bool success;
+    uint32_t featureMap;
+    success = (Attributes::FeatureMap::Get(endpoint, &featureMap) == EMBER_ZCL_STATUS_SUCCESS);
+
+    return success ? ((featureMap & to_underlying(feature)) != 0) : false;
+}
+
+/**********************************************************
+ * Callbacks Implementation
+ *********************************************************/
+
+bool emberAfTemperatureControlClusterSetTemperatureCallback(app::CommandHandler * commandObj,
+                                                            const app::ConcreteCommandPath & commandPath,
+                                                            const Commands::SetTemperature::DecodableType & commandData)
+{
+    auto & targetTemperature      = commandData.targetTemperature;
+    auto & targetTemperatureLevel = commandData.targetTemperatureLevel;
+    EndpointId endpoint           = commandPath.mEndpointId;
+    Status status                 = Status::Success;
+
+    if (TemperatureControlHasFeature(endpoint, TemperatureControl::Feature::kTemperatureNumber))
+    {
+        if (!targetTemperature.HasValue())
+        {
+            status = Status::Failure;
```

Comment:
This is not the right status.  Please follow the spec here.

Reply:
Updated code